diff --git a/cmake/modules/AddOCaml.cmake b/cmake/modules/AddOCaml.cmake
index 554046b20ed..31b711e1d21 100644
--- a/cmake/modules/AddOCaml.cmake
+++ b/cmake/modules/AddOCaml.cmake
@@ -23,6 +23,10 @@
 #

 function(add_ocaml_library name)
+  if( APPLE )
+    set (CMAKE_SHARED_LIBRARY_SUFFIX ".so")
+  endif()
+
   CMAKE_PARSE_ARGUMENTS(ARG "NOCOPY" "" "OCAML;OCAMLDEP;C;CFLAGS;PKG;LLVM" ${ARGN})

   set(src ${CMAKE_CURRENT_SOURCE_DIR})
@@ -53,9 +57,16 @@ function(add_ocaml_library name)
   endif()

   set(ocaml_flags "-lstdc++" "-ldopt" "-L${LLVM_LIBRARY_DIR}"
-                  "-ccopt" "-L\\$CAMLORIGIN/../.."
-                  "-ccopt" "-Wl,-rpath,\\$CAMLORIGIN/../.."
-                  ${ocaml_pkgs})
+                  "-ccopt" "-L\\$CAMLORIGIN/../..")
+
+  if (LLVM_OCAML_OUT_OF_TREE AND NOT LLVM_LINK_LLVM_DYLIB)
+    list(APPEND ocaml_flags "-ccopt" "-Wl,-rpath,${LLVM_OCAML_EXTERNAL_LLVM_LIBDIR}")
+    message (STATUS "Out of tree build with separate dylibs, set static rpath")
+  else()
+    list(APPEND ocaml_flags "-ccopt" "-Wl,-rpath,\\$CAMLORIGIN/../..")
+  endif()
+
+  list(APPEND ocaml_flags ${ocaml_pkgs})

   foreach( ocaml_dep ${ARG_OCAMLDEP} )
     get_target_property(dep_ocaml_flags "ocaml_${ocaml_dep}" OCAML_FLAGS)
@@ -74,6 +85,10 @@ function(add_ocaml_library name)
       list(APPEND ocaml_flags "-l${llvm_lib}" )
     endforeach()

+    if(LLVM_OCAML_OUT_OF_TREE)
+      list(APPEND ocaml_flags "-L${LLVM_OCAML_EXTERNAL_LLVM_LIBDIR}" )
+    endif()
+
     get_property(system_libs TARGET LLVMSupport PROPERTY LLVM_SYSTEM_LIBS)
     foreach(system_lib ${system_libs})
       if (system_lib MATCHES "^-")
@@ -148,7 +163,7 @@ function(add_ocaml_library name)
   elseif( UNIX )
     set(ocaml_rpath "\\$ORIGIN/../../../lib${LLVM_LIBDIR_SUFFIX}")
   endif()
-  list(APPEND ocaml_flags "-ldopt" "-Wl,-rpath,${ocaml_rpath}")
+  list(APPEND ocaml_flags "-ldopt" "-Wl,-rpath,${ocaml_rpath}" "-ldopt" "-Wl,-rpath,${LLVM_OCAML_EXTERNAL_LLVM_LIBDIR}")

   add_custom_command(
     OUTPUT ${ocaml_outputs}
